### ISSUE TITLE
Send players from one server to an other server.

### ISSUE THREAD

**Issue by Lejo1** [minetest/minetest#5813]
There was a forum topic about this: https://forum.minetest.net/viewtopic.php?f=5&t=17580

We should make server networks possible.
```Lua
-- Whether to enable or disable the whole server
-- network functionality (i.e. "be part of a network").
-- This implies that the whole authentication and
-- local logon functionality is disabled.
server_network_enabled = true

-- Define the authorative master server for sharing
-- the authentication information and session token
-- for individual users so they do not need to log on
-- manually on each server
server_network_master = master.example.com:30000
```
There should be a master server:
```Lua
-- Enable network and master server functionality
server_network_enabled = true
server_network_master = true -- defines THIS server as master server

-- The name of the network, shared between the devices
server_network_name = "My cool server network"

-- List of servers in the network (all user transfer requests
-- from servers not in this list will be ignored)
server_network_servers = server.example.com:30000, server2.example.net:30000, ..., ...

-- Whether or not share the inventory/chat between the
-- servers of the network
server_network_share_inventory = true
server_network_network_wide_chat = true

-- Allow porting to the set home point between the servers.
-- If disabled /home and /sethome are in relation to the server
-- a client is connected to.
server_network_allow_other_server_home = true

-- Allow join from client
allow_join_from_client = true
```
For sending players something like:
`minetest.send_player_to_server(player, address)`

EDIT: Fixed the forum link and moved the code into formatted area

**Comment by nerzhul** [comment]
This is an interesting idea but requires servers to have common auth, first and second a sort or cross server communication

**Comment by Lejo1** [comment]
of course it requires Servers, but even for little networks is this very nice.

**Comment by bigfoot547** [comment]
I would not hop between servers per se, but rather between ports.

**Comment by nerzhul** [comment]
@bigfoot547 from the point of view of a IP socket, address and port are required.

**Comment by dsohler** [comment]
You could hop from example.com:30000 to example.com:30001 when registering the second one as part of the network. There would be no difference between "other port on same hostname" and other machines.

Authentication and user metadata would be done completely by the server that acts as master server in the set up server network. No need for central auth servers.

**Comment by nerzhul** [comment]
as a en user you see that, as a developper i see socket(in_addr, port) :D

**Comment by nerzhul** [comment]
and no, authentication should be delegated to another service, but it's not the purpose of minetest ATM, we will rewrite the whole mientest transport protocol i think before that

**Comment by dsohler** [comment]
> we will rewrite the whole mientest transport protocol i think before that

I'm a big fan of rewriting stuff for modernizing it even if it causes backwards **in**compatibility :smile:

**Comment by benrob0329** [comment]
Why not just a simple function like ``player:send_to_server(server, port)`` and let the mods use http/s for inter server communication. Clients would default to using same uname/pword, opting to have a popup if auth failed.

**Comment by MuhdNurHidayat** [comment]
That function might be useful for server operator which runs multiple server on different modes and the players could go on the different servers directly without needing to quit the server and join other server from main menu.

I do saw a few server operators running servers like that, creative-only on one server, PvP on other server, survival on yet other server.

**Comment by nerzhul** [comment]
@benrob0329 it's more complicated than that, if we do that a server should generate a token to the other server which server accepts or not, and if accepted, player is redirected, password should not be kept in memory (and server doesn't know the password)

**Comment by Lejo1** [comment]
We can also say that players on servers of the server network don't have any password only name but then joining must only be allowed from servers out of the server network.
So set at all servers (not master) allow_join_from_client = false

**Comment by dsohler** [comment]
> We can also say that players on servers of the server network don't have any password 

No, they of course have a password. But the password is only managed by the master server .Thje other servers ask the master server if a user wants to connect. The master server then either provides the password or uses an external service for authenticateion and sends the result to the  "slave server" a user wants to connect to.

Sending users from one server to another has nothing to do with passwords anymore because the auth token already was created.

**Comment by benrob0329** [comment]
Imo inventory etc sync can be handled by mods via the http api, lets just get basic

1) send player token to other server and client
2) send client to other server
3) if client token matches token associated with username, allow joining
4) mods can set inv and other things

**Comment by nerzhul** [comment]
you talk about the most complicated thing ever for a monolitic program: distributed objects, it's not simple and security should be taken in consideration, very very high priority

**Comment by dsohler** [comment]
> Or even worse, people might suggest that minetest use a central authentication server

No, that would be stupid. Only the proposed "server networks" should have a central authentication server (or external authentication provider).

**Comment by benrob0329** [comment]
@nerzhul I don't talk of distributed objects, I talk of client redirection.

**Comment by Lejo1** [comment]
We can first also add just the minetest.send_player_to_server(player, address) Method and the player have to type in his name and password and if the server is down the player just get kicked.

**Comment by Distributed-DeadLock** [comment]
I*m in full support of every possible solution.
Why not implement a version with minimal functionality first, and work on the complicated details later?

**Comment by Calinou** [comment]
> I'm in full support of every possible solution.
> Why not implement a version with minimal functionality first, and work on the complicated details later?

Please do not bump issues without bringing significant new information.

**Comment by HeadClot** [comment]
Hey just thought I would drop this here
https://github.com/uetopia/servertravel-persistence-demo

It is in C++11 and it is for Unreal Engine 4. However it should give an Idea on how to do something like connecting from one server to another. :)

**Comment by Lejo1** [comment]
https://github.com/andrewvy/knot

**Comment by orwell96** [comment]
Those ideas of server networks sound interesting, but are far from what is achievable in the near future.
The only thing that is really necessary at the moment is a way to instruct a client to disconnect from one server and immediately connect to another server using the same username and password he used before. All other things can be implemented by other means:
- shared auth backend via postgresql database is planned in near future, and can already be achieved by a custom auth handler
- sharing position/homes is useless, because most probably the worlds will differ
- shared chat can be achieved by establishing some external bridge, like a slightly modified IRC mod
- inventory synchronisation is the most tricky part and will require some external webserver that ensures that a client is only logged in on one of the servers. The inventory would be backed up when the player disconnects and in regular intervals (synchronous to saving the player files in the world) to this webservice, and loaded from there when he joins a server (while completely ignoring the data written in the minetest player database)

**Comment by Lejo1** [comment]
I agree.
Of course a server network is not the first thing, that should be done.
minetest.transfer_player(target_address, target_port)
is a good beginning.

**Comment by niansa** [comment]
I like this idea! :D I hope it gets continue in a good way.

But... Why it doesn't work like this: Client connects to Server1.com:30000 . Server1.com says please connect to Anotherserver2.org:30000 (Optional: and gives the client a token. without the token the client cannot connect to Anotherserver.org:30000 ). The client disconnects and directly starts connection to Anotherserver2.org:30000 with the same username/password which the client used to connect to Server1.com (Optional: and sends the token which the client has got from Server1.com ).

**Comment by Lejo1** [comment]
Yeah!
The Subworlds has also been discussed here #4428

**Comment by paramat** [comment]
Any core dev support? Issue seems in a similar position to #4428

**Comment by paramat** [comment]
Discussion can continue in closely related #4428

**Comment by rubenwardy** [comment]
I support doing this. Can either reuse the password or show a dialog. As the server initiates, I don't think the former is a problem - the latter may condition the user into entering their password into dialogs that pop up

**Comment by rubenwardy** [comment]
I'd allow passing arbitrary Lua tables with the transfer request, the receiving server can then do as they wish. But yeah, you'd need to sign it as well to prevent tampering. This could be a later addition

**Comment by sfan5** [comment]
I'd rather let modders pass a string, that disincentivizes passing Lua tables (which need to be secured) at least a little bit.
**Or** if the engine allows passing Lua tables it should also provide the means to secure them (automatic signing & verification).

**Comment by nerzhul** [comment]
What about the inventory and items which can bé différent ? Le 11 oct. 2023 13:01, sfan5 ***@***.***> a écrit :
I'd rather let modders pass a string, that disincentivizes passing Lua tables (which need to be secured) at least a little bit.
Or if the engine allows passing Lua tables it should also provide the means to secure them (automatic signing & verification).

**Comment by rubenwardy** [comment]
> I'd rather let modders pass a string, that disincentivizes passing Lua tables (which need to be secured) at least a little bit. That's almost guaranteed to turn into a security nightname if you let modders do it. **Or** if the engine allows passing Lua tables it should also provide the means to secure them (automatic signing & verification).

I'd serialise the table as JSON to avoid using core.deserialize. I don't think this makes a difference to signing/verification though

**Comment by sfan5** [comment]
> What about the inventory and items which can bé différent ?

Nothing. If mods want to synchronize them they have to figure something out.
